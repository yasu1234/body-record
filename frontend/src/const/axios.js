import axios from "axios";
import Cookies from "js-cookie";

const baseURL = import.meta.env.VITE_APP_API_BASE;

const axiosInstance = axios.create({
  baseURL,
  headers: {
    "Content-Type": "application/json",
  },
});

axiosInstance.interceptors.request.use((config) => {
  config.headers["access-token"] = Cookies.get("accessToken");
  config.headers["client"] = Cookies.get("client");
  config.headers["uid"] = Cookies.get("uid");

  return config;
});

export default axiosInstance;
