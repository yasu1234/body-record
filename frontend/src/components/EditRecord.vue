<script setup>
import { ref, onMounted } from 'vue';
import axios from 'axios';
import { useRoute } from 'vue-router'
import Cookies from 'js-cookie';

import DropFile from './DropFile.vue'
import DatePicker from './DatePicker.vue'

const route = useRoute();

const recordDate = ref("");
const memo = ref("");
const files = ref([]);
const imageUrls = ref([]);
const recordId = ref(null);

onMounted(() => {
    getDetail();
});

function dateChange(event) {
    recordDate.value = event
}

function onFileChange(event) {
    files.value = [...event];
}

const deleteImage = async (item) => {
    try {
        const res = await axios.delete(import.meta.env.VITE_APP_API_BASE + '/api/v1/record/image', {
            params: {
                'id': recordId.value,
                'image_id': item.id
            },
            headers: {
                'access-token' : Cookies.get('accessToken'),
                'client':Cookies.get('client'),
                'uid': Cookies.get('uid')
            }
        })
        imageUrls.value = res.data.imageUrls
    } catch (error) {
        console.log({ error })
    }
}

const getDetail = async () => {
    const id = route.params.id
    try {
        const res = await axios.get(import.meta.env.VITE_APP_API_BASE + `/api/v1/records/${id}`, {
            headers: {
                'access-token' : Cookies.get('accessToken'),
                'client':Cookies.get('client'),
                'uid': Cookies.get('uid')
            }
        })
        recordId.value = res.data.record.id
        recordDate.value = res.data.record.date
        memo.value = res.data.record.memo
        imageUrls.value = res.data.imageUrls
    } catch (error) {
        console.log({ error })
    }
}

const edit = async () => {
    try {
        const formData = new FormData();
        formData.append('memo', memo.value);
        formData.append('date', recordDate.value);

        for (const file of files.value) {
            formData.append('images', file);
        }

        const res = await axios.patch(import.meta.env.VITE_APP_API_BASE + `/api/v1/records/${recordId.value}`, formData, {
            headers: {
                'Content-Type': 'multipart/form-data',
                'access-token' : Cookies.get('accessToken'),
                'client':Cookies.get('client'),
                'uid': Cookies.get('uid')
            }
        })
        recordId.value = res.data.record.id
        memo.value = res.data.record.memo
        recordDate.value = res.data.record.date
        imageUrls.value = res.data.imageUrls
    } catch (error) {
        console.log({ error })
    }
}
</script>

<template>
    <div class="edit-time">
        <p class="inputTitle">開始日</p>
        <DatePicker isStart=true :date= recordDate @update:date="dateChange"/>
    </div>
    <div class="editor">
        <textarea name="memo" rows="10" v-model="memo"></textarea>
    </div>
    <div class="thumbnail-container">
        <div class="thumbnail" v-for="item in imageUrls">
            <div class="thumbnail-image">
                <img :src="item.url" alt="">
            </div>
            <div class="thumbnail-actions">
                <button class="delete-button" @click="deleteImage(item)">X</button>
            </div>
        </div>
    </div>
    <div class="relationImages">
        <p class="inputTitle">関連画像</p>
        <DropFile @change="onFileChange"/>
    </div>
    <div class="relationImages">
        <button class="registerButton" @click="edit">編集する</button>
    </div>
</template>

<style>
.edit-time {
    padding-left: 20px;
    padding-right: 20px;
}

.editor{
   padding: 30px;
 }

 input[type=text] {
  width: 100%;
  padding: 12px 12px;
  margin: 8px 0;
  box-sizing: border-box;
  border: 1px solid #ccc;
  border-radius: 4px;
}
.time-list {
  display: flex;
  justify-content: space-between;
  padding-left: 30px;
  padding-right: 30px;
}
.time-list .item {
  width: 50%;
  margin: 0;
  padding: 10px;
  box-sizing: border-box;
}
.time-list .item .inputTitle {
  margin: 5px 0 0;
  padding: 0;
  font-size: 16px;
}
.relationImages {
    padding: 20px;
}
.registerButton{
    background: #ffa500;
    color: white;
    font-size:16px;
    font-weight:bold;
}
.thumbnail-container {
    display: flex;
    flex-wrap: wrap;
    justify-content: flex-start;
    padding: 20px;
}
.thumbnail {
    position: relative;
    display: inline-block;
    height: 200px;
    margin-right: 15px;
    margin-bottom: 15px;
    padding-left: 20px;
}
.thumbnail img {
    height: 100%;
}
.thumbnail-image {
    height: 100%;
}
.thumbnail-image img {
    height: 100%;
}
.thumbnail-actions {
    position: absolute;
    top: 5px;
    right: 5px;
}
.delete-button {
    background-color: rgba(255, 255, 255, 0.7);
    border: none;
    color: #000;
    padding: 5px 10px;
    text-align: center;
    text-decoration: none;
    display: inline-block;
    font-size: 12px;
    border-radius: 4px;
    cursor: pointer;
    border-radius: 50%
  }
</style>