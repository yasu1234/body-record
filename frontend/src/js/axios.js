import axios from "axios";
import Cookies from "js-cookie";

const baseURL = import.meta.env.VITE_APP_API_BASE;

const axiosInstance = axios.create({
  baseURL,
  headers: {
    "Content-Type": "application/json",
  },
});

axiosInstance.interceptors.request.use((config) => {
  config.headers["access-token"] = Cookies.get("accessToken");
  config.headers["client"] = Cookies.get("client");
  config.headers["uid"] = Cookies.get("uid");

  return config;
});

const setupInterceptors = (router) => {
  axiosInstance.interceptors.response.use(
    response => response,
    error => {
      // 共通のエラー処理
      if (error.response.status === 404) {
        router.push({ name: 'NotFound' })
      }

      // エラーをキャッチしてreturnする
      return Promise.reject(error)
    }
  )
}

export { axiosInstance, setupInterceptors }
