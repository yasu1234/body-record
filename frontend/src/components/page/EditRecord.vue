<script setup>
import { ref, onMounted } from 'vue';
import axios from 'axios';
import { useRoute, useRouter } from 'vue-router'
import Cookies from 'js-cookie';
import Toast from "primevue/toast";
import { useToast } from "primevue/usetoast";
import { toastService } from "../../const/toast.js";

import DropFile from '../atom/DropFile.vue'
import DatePicker from '../atom/DatePicker.vue'
import Header from '../layout/Header.vue'

const route = useRoute();
const router = useRouter();
const toast = useToast();
const toastNotifications = new toastService(toast);

const recordDate = ref("");
const memo = ref("");
const weight = ref(null);
const fatPercentage = ref(null);
const files = ref([]);
const imageUrls = ref([]);
const recordId = ref(null);

onMounted(() => {
    getDetail();
});

function dateChange(event) {
    recordDate.value = event
}

function onFileChange(event) {
    files.value = [...event];
}

const showNotFound = () =>  {
    router.push({ name: 'NotFound'})
}

const deleteImage = async (item) => {
    try {
        const res = await axios.delete(import.meta.env.VITE_APP_API_BASE + '/api/v1/record/image', {
            params: {
                'id': recordId.value,
                'image_id': item.id
            },
            headers: {
                'access-token' : Cookies.get('accessToken'),
                'client':Cookies.get('client'),
                'uid': Cookies.get('uid')
            }
        })
        imageUrls.value = res.data.imageUrls
    } catch (error) {
        console.log({ error })
    }
}

const getDetail = async () => {
    const id = route.params.id
    try {
        const res = await axios.get(import.meta.env.VITE_APP_API_BASE + `/api/v1/records/${id}`, {
            headers: {
                'access-token' : Cookies.get('accessToken'),
                'client':Cookies.get('client'),
                'uid': Cookies.get('uid')
            }
        })
        recordId.value = res.data.record.id
        recordDate.value = res.data.record.date
        memo.value = res.data.record.memo
        imageUrls.value = res.data.imageUrls
    } catch (error) {
        if (error.response.status === 404) {
            showNotFound()
        }
    }
}

const edit = async () => {
    try {
        const formData = new FormData();
        formData.append('record[memo]', memo.value);
        formData.append('record[date]', recordDate.value);
        if (weight.value !== null) {
            formData.append('record[weight]', weight.value);
        }
        if (fatPercentage.value !== null) {
            formData.append('record[fat_percentage]', fatPercentage.value);
        }

        for (const file of files.value) {
            formData.append('record[images]', file);
        }

        const res = await axios.patch(import.meta.env.VITE_APP_API_BASE + `/api/v1/records/${recordId.value}`, formData, {
            headers: {
                'Content-Type': 'multipart/form-data',
                'access-token' : Cookies.get('accessToken'),
                'client':Cookies.get('client'),
                'uid': Cookies.get('uid')
            }
        })
        recordId.value = res.data.record.id
        memo.value = res.data.record.memo
        recordDate.value = res.data.record.date
        imageUrls.value = res.data.imageUrls
    } catch (error) {
        if (error.response.status === 404) {
            showNotFound()
        } else {
            let errorMessages = "";
            if (error.response.status === 422) {
                if (Array.isArray(error.response.data.errors)) {
                    errorMessages += error.response.data.errors.join('\n');
                }
            }
            toastNotifications.displayError("記録の更新に失敗しました", errorMessages)
        }
    }
}
</script>

<template>
    <Header />
    <Toast position="top-center" />
    <div class="edit-time">
        <p class="inputTitle">開始日</p>
        <DatePicker isStart=true :date= recordDate @update:date="dateChange"/>
    </div>
    <div class="profile-edit-content">
        <label for="goal-weight">体重:</label>
        <input type="text" id="goal-weight" v-model="weight">
        <label for="goal-weight" class="unit-label">kg</label>
    </div>
    <div class="profile-edit-content">
        <label for="goal-fat-percentage">体脂肪率:</label>
        <input type="text" v-model="fatPercentage">
        <label for="goal-fat-percentage" class="unit-label">%</label>
    </div>
    <div class="editor">
        <textarea name="memo" rows="10" v-model="memo"></textarea>
    </div>
    <div class="thumbnail-container">
        <div class="thumbnail" v-for="item in imageUrls">
            <div class="thumbnail-image">
                <img :src="item.url" alt="">
            </div>
            <div class="thumbnail-actions">
                <button class="delete-button" @click="deleteImage(item)">X</button>
            </div>
        </div>
    </div>
    <div class="relationImages">
        <p class="inputTitle">関連画像</p>
        <DropFile @change="onFileChange"/>
    </div>
    <div class="relationImages">
        <button class="registerButton" @click="edit">編集する</button>
    </div>
</template>

<style scoped>
.edit-time {
    padding-left: 20px;
    padding-right: 20px;
}

.editor{
   padding: 30px;
 }

 input[type=text] {
  width: 100%;
  padding: 12px 12px;
  margin: 8px 0;
  box-sizing: border-box;
  border: 1px solid #ccc;
  border-radius: 4px;
}
.time-list {
  display: flex;
  justify-content: space-between;
  padding-left: 30px;
  padding-right: 30px;
}
.time-list .item {
  width: 50%;
  margin: 0;
  padding: 10px;
  box-sizing: border-box;
}
.time-list .item .inputTitle {
  margin: 5px 0 0;
  padding: 0;
  font-size: 16px;
}
.relationImages {
    padding: 20px;
}
.registerButton{
    background: #ffa500;
    color: white;
    font-size:16px;
    font-weight:bold;
}
.thumbnail-container {
    display: flex;
    flex-wrap: wrap;
    justify-content: flex-start;
    padding: 20px;
}
.thumbnail {
    position: relative;
    display: inline-block;
    height: 200px;
    margin-right: 15px;
    margin-bottom: 15px;
    padding-left: 20px;
}
.thumbnail img {
    height: 100%;
}
.thumbnail-image {
    height: 100%;
}
.thumbnail-image img {
    height: 100%;
}
.thumbnail-actions {
    position: absolute;
    top: 5px;
    right: 5px;
}
.delete-button {
    background-color: rgba(255, 255, 255, 0.7);
    border: none;
    color: #000;
    padding: 5px 10px;
    text-align: center;
    text-decoration: none;
    display: inline-block;
    font-size: 12px;
    border-radius: 4px;
    cursor: pointer;
    border-radius: 50%
  }
</style>
