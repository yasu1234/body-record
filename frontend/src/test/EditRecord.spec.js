import { shallowMount, flushPromises } from '@vue/test-utils'
import axios from 'axios'
import EditRecord from '@/components/page/EditRecord.vue'

jest.mock('axios')
jest.mock('vue-router', () => {
    const router = {
      push: jest.fn(),
      pathname: '/',
      route: '',
      asPath: '',
      query: {},
      events: {
        on: jest.fn(),
        off: jest.fn(),
        emit: jest.fn(),
      },
    };
    return {
      useRouter: jest.fn().mockReturnValue(router),
      useRoute: jest.fn().mockReturnValue({
        params: {
          id: 1
        }
      })
    };
  });

  jest.mock('primevue/usetoast', () => ({
    useToast: jest.fn(() => ({
      add: jest.fn(),
      removeAll: jest.fn()
    }))
  }))

describe('EditRecord', () => {
  let wrapper

  const mockRecord = {
    id: 1,
    title: 'Test',
    image_url: [
      { id: 1, url: 'https://example.com/image1.jpg' },
      { id: 2, url: 'https://example.com/image2.jpg' }
    ]
  } 

  // onMounted内でaxios.getが呼ばれるため、このタイミングでモック設定
  axios.get.mockResolvedValue({ data: { record: mockRecord } })

  beforeEach(() => {
    wrapper = shallowMount(EditRecord, {
        global: {
          stubs: ["router-link", "router-view"]
        }
    })
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('SnapShotテスト初期表示', () => {
    expect(wrapper.element).toMatchSnapshot();
  })
})
