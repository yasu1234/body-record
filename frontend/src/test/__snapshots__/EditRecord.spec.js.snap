// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditRecord SnapShotテスト初期表示 1`] = `
<div
  data-v-app=""
>
  
  <header-stub />
  <tab-menu-stub />
  <toast-stub
    position="top-center"
  />
  <form
    class="edit-container"
  >
    <div
      class="w-52"
    >
      <p>
        記録日
      </p>
      <date-picker-stub
        isstart="true"
      />
    </div>
    <div
      class="weight-group mt-7"
    >
      <float-label-stub />
      <p
        class="ml-2"
      >
        kg
      </p>
    </div>
    <div
      class="weight-group mt-7"
    >
      <float-label-stub />
      <p
        class="ml-2"
      >
        %
      </p>
    </div>
    <div
      class="mt-5"
    >
      <record-memo-input-stub />
    </div>
    <div
      class="mt-7"
    >
      <input
        class="w-4 h-4 text-blue-600 bg-gray-100 border-gray-300 rounded focus:ring-blue-500"
        id="statusSelect"
        type="checkbox"
      />
      <label
        class="ml-2"
      >
        非公開記録にする場合にはチェック
      </label>
    </div>
    <div>
      <p
        class="mt-5"
      >
        登録済みの画像
      </p>
      <div
        class="thumbnail-container"
      >
        
        <div
          class="thumbnail"
        >
          <div
            class="thumbnail-image"
          >
            <img
              alt=""
              src="https://example.com/image1.jpg"
            />
          </div>
          <div
            class="thumbnail-actions"
          >
            <button-stub
              class="delete-button"
              icon="pi pi-trash"
              label=""
            />
          </div>
        </div>
        <div
          class="thumbnail"
        >
          <div
            class="thumbnail-image"
          >
            <img
              alt=""
              src="https://example.com/image2.jpg"
            />
          </div>
          <div
            class="thumbnail-actions"
          >
            <button-stub
              class="delete-button"
              icon="pi pi-trash"
              label=""
            />
          </div>
        </div>
        
      </div>
    </div>
    <div
      class="mt-5"
    >
      <p>
        関連画像(登録済みの画像と合わせて3枚まで登録できます)
      </p>
      
      <div>
        <drop-file-stub
          class="mt-3"
          index="1"
        />
      </div>
      <div>
        <drop-file-stub
          class="mt-3"
          index="2"
        />
      </div>
      <div>
        <drop-file-stub
          class="mt-3"
          index="3"
        />
      </div>
      
    </div>
    <div
      class="p-5 text-center"
    >
      <button
        class="record-edit-button"
      >
        編集する
      </button>
    </div>
  </form>
  
</div>
`;
